proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,211 @@
+        | delegate{l} instr* end | caught_m{a val^n} instr* end
+```
+
+Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.
+
+```
+label_kind ::= try_label_kind | catch_label_kind
+try_label_kind ::= catch_m{a? instr*}* | delegate{l}
+catch_label_kind ::= caught_m{a val*}
+B^0 ::= val* [_] instr*
+B^{k+1} ::= val* label_n{instr*} label_kind? B^k end? end instr*
+```
+
+Note that `label_n{instr*} label_kind? [_] end? end` could be seen as a simplified control frame.
+
+(Alternatively, we could have the above `label_kind`s be also labels,  remove the additional `label_m` from the execution rules below, and remove the execution rules below where the new administrative instructions only contain `val*`. This would make labels even more similar to control frames.)
+
+Throw Contexts
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end
+   | frame_n{F} T end
+```
+
+### Instructions
+
+
+```
+F; throw x  ↪  F; throw a  (if F.module.exnaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
+
+caught_m{a val^n} val^m end  ↪  val^m
+```
+
+A missing tagaddr in a `catch_m` clause (i.e., `a? = ε`) represents a `catch_all`.
+
+```
+F; val^n (try bt instr* (catch x instr'*)* (catch_all instr''*)? end)
+  ↪  F; label_m{} (catch_m{a instr'*}*{instr''*}? val^n instr* end) end
+  (if bt = [t1^n]→[t2^m] ∧ (val:t1)^n ∧ (F.module.exnaddrs[x]=a)*)
+
+catch_m{a? instr*}* val^m end ↪ val^m
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; caught_m{a val^n} (val^n)? instr* end
+  (if (a1? = ε ∨ a1? = a) ∧ S.tags(a).type = [t^n]→[])
+
+S; F; catch_m{a1? instr*}{a'? instr'*}* T[val^n (throw a)] end
+  ↪  S; F; catch_m{a'? instr'*}* T[val^n (throw a)] end
+  (if a1? ≠ ε ∧ a1? ≠ a)
+
+S; F; catch_m T[val^n (throw a)] end
+  ↪  S; F; val^n (throw a)
+
+
+val^n (try bt instr* delegate l)
+  ↪ label_m{} (delegate{l} val^n instr* end) end
+  (if bt = [t^n]→[t^m])
+
+delegate{l} val^n end ↪ val^n
+
+B^l[ delegate{l} T[val^n (throw a)] end ]
+  ↪ val^n (throw a)
+```
+
+### Typing rules for administrative instructions
+
+```
+S.tags[a].type = [t2*]→[]
+--------------------------------
+S;C ⊢ throw a : [t1* t2*]→[t*]
+
+(S.tags[a].type = [t'*]→[] ∧
+ S;C, labels {result [t*], kind catch} ⊢ instr1* : [t'*]→[t*])*
+(S;C, labels {result [t*], kind catch} ⊢ instr2* : []→[t*])?
+S;C, labels {result [t*], kind try} ⊢ instr3* : []→[t*]
+-----------------------------------------------------------------------
+S;C, labels [t*] ⊢ catch_n{a instr1*}*{instr2*}? instr3* end : []→[t*]
```

Comment:
This could be simplified somewhat:
```suggestion
((S.tags[a].type = [t'*]→[])?
 S;C, labels {result [t*], kind catch} ⊢ instr'* : [t'*?]→[t*])*
S;C, labels {result [t*], kind try} ⊢ instr* : []→[t*]
-----------------------------------------------------------------------
S;C, labels [t*] ⊢ catch_n{a? instr'*}* instr* end : []→[t*]
```